hpx/parallel/algorithms/merge.hpp

```diff
@@ -491,6 +490,341 @@ namespace hpx { namespace parallel { inline namespace v1
+                        std::current_exception());
+
+                    std::list<std::exception_ptr> errors;
+                    util::detail::handle_local_exceptions<ExPolicy>::call(
+                        futures, errors);
+
+                    // Not reachable.
+                    HPX_ASSERT(false);
+                }
+
+                fut.get();
+            }
+            else /* left_size < right_size */
+            {
+                // Means that always 'pivot' < 'last'.
+                HPX_ASSERT(right_size >= 3ul);
+
+                // Select pivot in rightside range.
+                RandIter pivot = middle + right_size / 2;
+                RandIter boundary = upper_bound_helper::call(
+                    first, middle, hpx::util::invoke(proj, *pivot), comp, proj);
+                RandIter target = boundary + (pivot - middle);
+
+                // Swap two blocks, [boundary, middle) and [middle, pivot+1).
+                // After this, [first, last) will be divided into three blocks,
+                //   [first, target), target, and [target+1, last).
+                // And all elements of [first, target) are less than
+                //   the thing of target.
+                // And all elements of [target+1, last) are greater or equal than
+                //   the thing of target.
+                std::rotate(boundary, middle, pivot + 1);
+
+                hpx::future<void> fut = execution::async_execute(
+                    policy.executor(),
+                    [&]() -> void
+                    {
+                        // Process the range which is leftside of 'target'.
+                        parallel_inplace_merge_helper(policy,
+                            first, boundary, target,
+                            comp, proj);
+                    });
+
+                try {
+                    // Process the range which is rightside of 'target'.
+                    parallel_inplace_merge_helper(policy,
+                        target + 1, pivot + 1, last,
+                        comp, proj);
+                }
+                catch (...) {
+                    fut.wait();
+
+                    std::vector<hpx::future<void>> futures(2);
+                    futures[0] = std::move(fut);
+                    futures[1] = hpx::make_exceptional_future<void>(
+                        std::current_exception());
+
+                    std::list<std::exception_ptr> errors;
+                    util::detail::handle_local_exceptions<ExPolicy>::call(
+                        futures, errors);
+
+                    // Not reachable.
+                    HPX_ASSERT(false);
+                }
+
+                fut.get();
+            }
+        }
+
+        template <typename ExPolicy, typename RandIter,
+            typename Comp, typename Proj,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_random_access_iterator<RandIter>::value)>
```

Comment:
@hkaiser Is this use of concepts strange, too?

Reply:
No need to change things.

Reply:
@hkaiser But, I removed that concept for consistency of codes.